File: Application/src/main/java/com/shareyourproxy/app/EditGroupChannelsActivity.java
package com.shareyourproxy.app;

import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.shareyourproxy.R;
import com.shareyourproxy.api.domain.model.Group;
import com.shareyourproxy.api.rx.JustObserver;
import com.shareyourproxy.api.rx.command.UpdateUserContactsCommand;
import com.shareyourproxy.api.rx.command.eventcallback.UserGroupDeletedEventCallback;
import com.shareyourproxy.api.rx.event.ViewGroupContactsEvent;
import com.shareyourproxy.app.fragment.AggregateFeedFragment;
import com.shareyourproxy.app.fragment.EditGroupChannelsFragment;

import java.util.ArrayList;

import butterknife.Bind;
import butterknife.ButterKnife;
import rx.subscriptions.CompositeSubscription;

import static com.shareyourproxy.Constants.ARG_EDIT_GROUP_TYPE;
import static com.shareyourproxy.Constants.ARG_SELECTED_GROUP;
import static com.shareyourproxy.IntentLauncher.launchEditGroupContactsActivity;
import static com.shareyourproxy.IntentLauncher.launchMainActivity;
import static com.shareyourproxy.app.EditGroupChannelsActivity.GroupEditType.ADD_GROUP;
import static com.shareyourproxy.app.EditGroupChannelsActivity.GroupEditType.EDIT_GROUP;
import static com.shareyourproxy.app.EditGroupChannelsActivity.GroupEditType.PUBLIC_GROUP;
import static com.shareyourproxy.util.ViewUtils.getMenuIconSecondary;
import static com.shareyourproxy.util.ViewUtils.hideSoftwareKeyboard;

/**
 * Add and remove newChannel permissions from a group.
 */
public class EditGroupChannelsActivity extends BaseActivity {
    @Bind(R.id.activity_toolbar)
    Toolbar toolbar;
    private CompositeSubscription _subscriptions;

    @Override
    public void onBackPressed() {
        hideSoftwareKeyboard(toolbar);
        finish();
        overridePendingTransition(R.anim.fade_in, R.anim.slide_out_bottom);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.common_activity_fragment_container);
        ButterKnife.bind(this);
        initialize();
        if (savedInstanceState == null) {
            getSupportFragmentManager().beginTransaction()
                .replace(R.id.activity_fragment_container,
                    EditGroupChannelsFragment.newInstance()).commit();
        }
    }

    private GroupEditType getAddOrEdit() {
        return (GroupEditType) getIntent().getExtras().getSerializable(ARG_EDIT_GROUP_TYPE);
    }

    private void initialize() {
        GroupEditType editType = getAddOrEdit();
        if (ADD_GROUP.equals(editType)) {
            buildToolbar(toolbar, getString(R.string.add_group),
                getMenuIconSecondary(this, R.raw.ic_clear));
        } else if (EDIT_GROUP.equals(editType)) {
            buildToolbar(toolbar, getString(R.string.edit_group),
                getMenuIconSecondary(this, R.raw.ic_clear));
        } else if (PUBLIC_GROUP.equals(editType)) {
            buildToolbar(toolbar, getString(R.string.public_group),
                getMenuIconSecondary(this, R.raw.ic_clear));
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        _subscriptions = new CompositeSubscription();
        _subscriptions.add(getRxBus().toObservable()
            .subscribe(onNextEvent(this)));
    }

    private JustObserver<Object> onNextEvent(final EditGroupChannelsActivity activity) {
        return new JustObserver<Object>() {
            @Override
            public void next(Object event) {
                if (event instanceof UserGroupDeletedEventCallback) {
                    userGroupDeleted((UserGroupDeletedEventCallback) event);
                } else if (event instanceof ViewGroupContactsEvent) {
                    launchEditGroupContactsActivity(activity, getSelectedGroup());
                }
            }
        };
    }

    private Group getSelectedGroup() {
        return getIntent().getExtras().getParcelable(ARG_SELECTED_GROUP);
    }

    @Override
    protected void onPause() {
        super.onPause();
        _subscriptions.unsubscribe();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_activity_edit_group_channel, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuItem search = menu.findItem(R.id.menu_edit_group_channel_save);
        search.setIcon(getMenuIconSecondary(this, R.raw.ic_done));
        return super.onPrepareOptionsMenu(menu);
    }

    private void userGroupDeleted(UserGroupDeletedEventCallback event) {
        updateUserContacts(event);
        launchMainActivity(this, AggregateFeedFragment.ARG_SELECT_GROUP_TAB, true, event.getGroup());
        onBackPressed();
    }

    private void updateUserContacts(UserGroupDeletedEventCallback event) {
        ArrayList<String> contacts = new ArrayList<>();
        if (event.getGroup().contacts() != null) {
            for (String contactId : event.getGroup().contacts()) {
                contacts.add(contactId);
            }
        }

        getRxBus().post(new UpdateUserContactsCommand(
            getLoggedInUser(), contacts, getLoggedInUser().groups()));
    }

    public enum GroupEditType {
        ADD_GROUP, EDIT_GROUP, PUBLIC_GROUP
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/app/EditGroupChannelsActivity.kt
package com.shareyourproxy.app

import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.view.Menu
import com.shareyourproxy.Constants.ARG_EDIT_GROUP_TYPE
import com.shareyourproxy.Constants.ARG_SELECTED_GROUP
import com.shareyourproxy.IntentLauncher.launchEditGroupContactsActivity
import com.shareyourproxy.IntentLauncher.launchMainActivity
import com.shareyourproxy.R
import com.shareyourproxy.R.anim.fade_in
import com.shareyourproxy.R.anim.slide_out_bottom
import com.shareyourproxy.R.id.activity_fragment_container
import com.shareyourproxy.R.id.menu_edit_group_channel_save
import com.shareyourproxy.R.layout.common_activity_fragment_container
import com.shareyourproxy.R.raw.ic_clear
import com.shareyourproxy.R.raw.ic_done
import com.shareyourproxy.R.string.*
import com.shareyourproxy.api.domain.model.Group
import com.shareyourproxy.api.rx.JustObserver
import com.shareyourproxy.api.rx.RxBusRelay.post
import com.shareyourproxy.api.rx.RxBusRelay.rxBusObservable
import com.shareyourproxy.api.rx.command.UpdateUserContactsCommand
import com.shareyourproxy.api.rx.command.eventcallback.UserGroupDeletedEventCallback
import com.shareyourproxy.api.rx.event.ViewGroupContactsEvent
import com.shareyourproxy.app.fragment.AggregateFeedFragment.Companion.ARG_SELECT_GROUP_TAB
import com.shareyourproxy.app.fragment.EditGroupChannelsFragment
import com.shareyourproxy.util.ButterKnife.LazyVal
import com.shareyourproxy.util.ButterKnife.bindView
import com.shareyourproxy.util.Enumerations.GroupEditType
import com.shareyourproxy.util.Enumerations.GroupEditType.*
import com.shareyourproxy.util.ViewUtils.getMenuIconSecondary
import com.shareyourproxy.util.ViewUtils.hideSoftwareKeyboard
import java.util.*

/**
 * Add and remove newChannel permissions from a group.
 */
private final class EditGroupChannelsActivity : BaseActivity() {
    private val toolbar: Toolbar by bindView(R.id.activity_toolbar)
    private val selectedGroup: Group by LazyVal { intent.extras.getParcelable<Group>(ARG_SELECTED_GROUP) }
    private val addOrEdit: GroupEditType by LazyVal { intent.extras.getSerializable(ARG_EDIT_GROUP_TYPE) as GroupEditType }

    override fun onBackPressed() {
        hideSoftwareKeyboard(toolbar)
        finish()
        overridePendingTransition(fade_in, slide_out_bottom)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(common_activity_fragment_container)
        initialize()
        if (savedInstanceState == null) {
            supportFragmentManager.beginTransaction().replace(activity_fragment_container, EditGroupChannelsFragment()).commit()
        }
    }

    override fun onResume() {
        super.onResume()
        rxBusObservable().subscribe(onNextEvent(this))
    }

    private fun initialize() {
        val editType = addOrEdit
        when (editType) {
            ADD_GROUP -> buildToolbar(toolbar, getString(add_group), getMenuIconSecondary(this, ic_clear))
            EDIT_GROUP -> buildToolbar(toolbar, getString(edit_group), getMenuIconSecondary(this, ic_clear))
            PUBLIC_GROUP -> buildToolbar(toolbar, getString(public_group), getMenuIconSecondary(this, ic_clear))
        }
    }

    private fun onNextEvent(activity: EditGroupChannelsActivity): JustObserver<Any> {
        return object : JustObserver<Any>(EditGroupChannelsActivity::class.java) {
            @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
            override fun next(event: Any) {
                when (event) {
                    is UserGroupDeletedEventCallback -> userGroupDeleted(event)
                    is ViewGroupContactsEvent -> launchEditGroupContactsActivity(activity, selectedGroup)
                }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_activity_edit_group_channel, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val search = menu.findItem(menu_edit_group_channel_save)
        search.setIcon(getMenuIconSecondary(this, ic_done))
        return super.onPrepareOptionsMenu(menu)
    }

    private fun userGroupDeleted(event: UserGroupDeletedEventCallback) {
        updateUserContacts(event)
        launchMainActivity(this, ARG_SELECT_GROUP_TAB, true, event.group)
        onBackPressed()
    }

    private fun updateUserContacts(event: UserGroupDeletedEventCallback) {
        val contacts = ArrayList<String>()
        event.group.contacts.forEach { contacts.add(it) }
        post(UpdateUserContactsCommand(loggedInUser, contacts, loggedInUser.groups))
    }
}
